### PR TITLE
BigQuery: Allow subset of schema to be passed into `load_table_from_dataframe`.

### LINKED ISSUE #8140: BigQuery: Allow partial schema in `load_table_from_dataframe`
As discussed in https://github.com/googleapis/google-cloud-python/pull/8105#discussion_r286997609 and implemented in pandas-gbq at https://github.com/pydata/pandas-gbq/issues/218, often times the default schema detection works well enough for pandas -> BQ, and only one or two of many columns need to be overridden.

Example:

```python
df = pandas.DataFrame({... lots of columns ..., "needs_manual_schema": [...]})
job_config = bigquery.LoadJobConfig(schema=[
    bigquery.SchemaField("needs_manual_schema", "BOOLEAN"),
])
client.load_table_from_dataframe(
    df, "my_dataset.my_table", job_config=job_config,
)
# All columns from df are uploaded, but only df["needs_manual_schema"] is
# coerced into a specific type (BOOLEAN).
```

### PR DESCRIPTION
**Pull request by tswast**
Closes #8140 

- [x] Allow subset of schema to be passed into `load_table_from_dataframe`.
- [x] The types of any remaining columns will be auto-detected.
- [x] When the type of any remaining column can't be determined, warn that the schema was ignored because the type of column X could not be determined.
- ~Update snippets to show overriding schema for columns whose types can't be autodetected.~ Blocked by https://github.com/googleapis/google-cloud-python/issues/5572#issuecomment-524012299.

~This PR is branched from https://github.com/googleapis/google-cloud-python/pull/9049.~

### CODE REVIEW

**Comment by tswast** [comment]
Blocked by https://github.com/googleapis/google-cloud-python/issues/9074

File: bigquery/tests/unit/test_client.py [hunk 1]
```diff
@@ -5230,7 +5230,7 @@ def test_load_table_from_dataframe(self):
         from google.cloud.bigquery import job
 
         client = self._make_client()
-        records = [{"name": "Monty", "age": 100}, {"name": "Python", "age": 60}]
+        records = [{"id": 1, "age": 100}, {"id": 2, "age": 60}]
```
Comment (plamut) [inline bigquery/tests/unit/test_client.py]: Any special reason for replacing a string column with an integer one?
(not that there's anything wrong with it, it looks fine, just being curious)

File: bigquery/tests/system.py [hunk 2]
```diff
@@ -847,44 +848,58 @@ def test_load_table_from_dataframe_w_explicit_schema(self):
...
+                ("int_col", [-9223372036854775808, None, 9223372036854775807]),
+                (
+                    "num_col",
+                    [
+                        decimal.Decimal("-99999999999999999999999999999.999999999"),
+                        None,
+                        decimal.Decimal("99999999999999999999999999999.999999999"),
+                    ],
+                ),
+                ("str_col", ["abc", None, "def"]),
```
Comment (plamut) [inline bigquery/tests/system.py]: Under Python 2, this is actually the same data as in `'bytes_col'`. It would make sense to use unicode literals instead, preferably with non-ASCII characters, as this is when things most often break.

(it's not strictly in the scope of this very commit, but if it can be easily done, let's harden the tests)

I noticed a few other places when normal string literals are used, please check if it makes sense to replace then with unicode literals (e.g. Pandas might recognize those as columns as bytes).

File: bigquery/google/cloud/bigquery/_pandas_helpers.py [hunk 3]
```diff
@@ -227,7 +227,7 @@ def dataframe_to_bq_schema(dataframe, bq_schema):
         bq_type = _PANDAS_DTYPE_TO_BQ.get(dtype.name)
         if not bq_type:
             warnings.warn("Unable to determine type of column '{}'.".format(column))
-            return ()
+            return None
```
Comment (plamut) [inline bigquery/google/cloud/bigquery/_pandas_helpers.py]: Is there a specific reason for reverting the return value back to `None`?

File: bigquery/tests/system.py [hunk 4]
```diff
@@ -847,44 +848,58 @@ def test_load_table_from_dataframe_w_explicit_schema(self):
...
+                    "time_col",
+                    [datetime.time(0, 0, 0), None, datetime.time(23, 59, 59, 999999)],
+                ),
+                (
+                    "ts_col",
+                    [
+                        datetime.datetime(1, 1, 1, 0, 0, 0, tzinfo=pytz.utc),
+                        None,
+                        datetime.datetime(
+                            9999, 12, 31, 23, 59, 59, 999999, tzinfo=pytz.utc
```
Comment (plamut) [inline bigquery/tests/system.py]: I like this and several other edge case values in test data. :+1:

File: bigquery/tests/unit/test_client.py (hunk 1 above)
Comment (tswast) [reply bigquery/tests/unit/test_client.py]: Since string columns have a pandas dtype of `object`, we end up hitting the soon-to-be-deprecated "no schema available" path. Alternatively, I could have used a partial schema in these tests, but that didn't seem relevant to the purpose of these tests.

Comment (plamut) [reply bigquery/tests/unit/test_client.py]: Makes sense, thanks.

File: bigquery/google/cloud/bigquery/_pandas_helpers.py (hunk 3 above)
Comment (tswast) [reply bigquery/google/cloud/bigquery/_pandas_helpers.py]: Since there is already a partial schema set, we have to unset it somehow. Otherwise, the number of columns in the load job does not match the number of columns in the file.

It turns out, setting an empty schema has a different behavior in the backend than not setting a schema at all.

```
___________________ test_load_table_from_dataframe[pyarrow] ____________________
client = <google.cloud.bigquery.client.Client object at 0x7f919ef73450>
to_delete = [Dataset(DatasetReference(u'precise-truck-742', 'load_table_from_dataframe_1566428806902'))]
parquet_engine = 'pyarrow'
    @pytest.mark.skipif(pandas is None, reason="Requires `pandas`")
    @pytest.mark.parametrize("parquet_engine", ["pyarrow", "fastparquet"])
    def test_load_table_from_dataframe(client, to_delete, parquet_engine):
        if parquet_engine == "pyarrow" and pyarrow is None:
            pytest.skip("Requires `pyarrow`")
        if parquet_engine == "fastparquet" and fastparquet is None:
            pytest.skip("Requires `fastparquet`")
        pandas.set_option("io.parquet.engine", parquet_engine)
        dataset_id = "load_table_from_dataframe_{}".format(_millis())
        dataset = bigquery.Dataset(client.dataset(dataset_id))
        client.create_dataset(dataset)
        to_delete.append(dataset)
        # [START bigquery_load_table_dataframe]
        # from google.cloud import bigquery
        # import pandas
        # client = bigquery.Client()
        # dataset_id = 'my_dataset'
        dataset_ref = client.dataset(dataset_id)
        table_ref = dataset_ref.table("monty_python")
        records = [
            {"title": u"The Meaning of Life", "release_year": 1983},
            {"title": u"Monty Python and the Holy Grail", "release_year": 1975},
            {"title": u"Life of Brian", "release_year": 1979},
            {"title": u"And Now for Something Completely Different", "release_year": 1971},
        ]
        # Optionally set explicit indices.
        # If indices are not specified, a column will be created for the default
        # indices created by pandas.
        index = [u"Q24980", u"Q25043", u"Q24953", u"Q16403"]
        dataframe = pandas.DataFrame(records, index=pandas.Index(index, name="wikidata_id"))
>       job = client.load_table_from_dataframe(dataframe, table_ref, location="US")
docs/snippets.py:2779:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
google/cloud/bigquery/client.py:1574: in load_table_from_dataframe
    job_config=job_config,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <google.cloud.bigquery.client.Client object at 0x7f919ef73450>
file_obj = <closed file '/tmp/tmpoL17bp_job_31ab0413.parquet', mode 'rb' at 0x7f919ef4a540>
destination = TableReference(DatasetReference(u'precise-truck-742', 'load_table_from_dataframe_1566428806902'), 'monty_python')
rewind = True, size = None, num_retries = 6
job_id = '31ab0413-37cd-4ff8-b417-446f356846fa', job_id_prefix = None
location = 'US', project = 'precise-truck-742'
job_config = <google.cloud.bigquery.job.LoadJobConfig object at 0x7f9194de2890>
    def load_table_from_file(
        self,
        file_obj,
        destination,
        rewind=False,
        size=None,
        num_retries=_DEFAULT_NUM_RETRIES,
        job_id=None,
        job_id_prefix=None,
        location=None,
        project=None,
        job_config=None,
    ):
        """Upload the contents of this table from a file-like object.
        Similar to :meth:`load_table_from_uri`, this method creates, starts and
        returns a :class:`~google.cloud.bigquery.job.LoadJob`.
        Arguments:
            file_obj (file): A file handle opened in binary mode for reading.
            destination (Union[ \
                :class:`~google.cloud.bigquery.table.Table`, \
                :class:`~google.cloud.bigquery.table.TableReference`, \
                str, \
            ]):
                Table into which data is to be loaded. If a string is passed
                in, this method attempts to create a table reference from a
                string using
                :func:`google.cloud.bigquery.table.TableReference.from_string`.
        Keyword Arguments:
            rewind (bool):
                If True, seek to the beginning of the file handle before
                reading the file.
            size (int):
                The number of bytes to read from the file handle. If size is
                ``None`` or large, resumable upload will be used. Otherwise,
                multipart upload will be used.
            num_retries (int): Number of upload retries. Defaults to 6.
            job_id (str): (Optional) Name of the job.
            job_id_prefix (str):
                (Optional) the user-provided prefix for a randomly generated
                job ID. This parameter will be ignored if a ``job_id`` is
                also given.
            location (str):
                Location where to run the job. Must match the location of the
                destination table.
            project (str):
                Project ID of the project of where to run the job. Defaults
                to the client's project.
            job_config (google.cloud.bigquery.job.LoadJobConfig):
                (Optional) Extra configuration options for the job.
        Returns:
            google.cloud.bigquery.job.LoadJob: A new load job.
        Raises:
            ValueError:
                If ``size`` is not passed in and can not be determined, or if
                the ``file_obj`` can be detected to be a file opened in text
                mode.
        """
        job_id = _make_job_id(job_id, job_id_prefix)
        if project is None:
            project = self.project
        if location is None:
            location = self.location
        destination = _table_arg_to_table_ref(destination, default_project=self.project)
        job_ref = job._JobReference(job_id, project=project, location=location)
        load_job = job.LoadJob(job_ref, None, destination, self, job_config)
        job_resource = load_job.to_api_repr()
        if rewind:
            file_obj.seek(0, os.SEEK_SET)
        _check_mode(file_obj)
        try:
            if size is None or size >= _MAX_MULTIPART_SIZE:
                response = self._do_resumable_upload(
                    file_obj, job_resource, num_retries
                )
            else:
                response = self._do_multipart_upload(
                    file_obj, job_resource, size, num_retries
                )
        except resumable_media.InvalidResponse as exc:
>           raise exceptions.from_http_response(exc.response)
E           BadRequest: 400 POST https://www.googleapis.com/upload/bigquery/v2/projects/precise-truck-742/jobs?uploadType=resumable: Empty schema specified for the load job. Please specify a schema that describes the data being loaded.
google/cloud/bigquery/client.py:1439: BadRequest
=============================== warnings summary ===============================
docs/snippets.py::test_load_table_from_dataframe[pyarrow]
  /tmpfs/src/github/google-cloud-python/bigquery/google/cloud/bigquery/_pandas_helpers.py:229: UserWarning: Unable to determine type of column 'title'.
    warnings.warn("Unable to determine type of column '{}'.".format(column))
```

To fix this, the correct thing to do is to unset the schema if set, which is why I filed: https://github.com/googleapis/google-cloud-python/issues/9074

```
=================================== FAILURES ===================================
___________________ test_load_table_from_dataframe[pyarrow] ____________________
client = <google.cloud.bigquery.client.Client object at 0x7f9963974450>
to_delete = [Dataset(DatasetReference(u'precise-truck-742', 'load_table_from_dataframe_1566434996535'))]
parquet_engine = 'pyarrow'
    @pytest.mark.skipif(pandas is None, reason="Requires `pandas`")
    @pytest.mark.parametrize("parquet_engine", ["pyarrow", "fastparquet"])
    def test_load_table_from_dataframe(client, to_delete, parquet_engine):
        if parquet_engine == "pyarrow" and pyarrow is None:
            pytest.skip("Requires `pyarrow`")
        if parquet_engine == "fastparquet" and fastparquet is None:
            pytest.skip("Requires `fastparquet`")
        pandas.set_option("io.parquet.engine", parquet_engine)
        dataset_id = "load_table_from_dataframe_{}".format(_millis())
        dataset = bigquery.Dataset(client.dataset(dataset_id))
        client.create_dataset(dataset)
        to_delete.append(dataset)
        # [START bigquery_load_table_dataframe]
        # from google.cloud import bigquery
        # import pandas
        # client = bigquery.Client()
        # dataset_id = 'my_dataset'
        dataset_ref = client.dataset(dataset_id)
        table_ref = dataset_ref.table("monty_python")
        records = [
            {"title": u"The Meaning of Life", "release_year": 1983},
            {"title": u"Monty Python and the Holy Grail", "release_year": 1975},
            {"title": u"Life of Brian", "release_year": 1979},
            {"title": u"And Now for Something Completely Different", "release_year": 1971},
        ]
        # Optionally set explicit indices.
        # If indices are not specified, a column will be created for the default
        # indices created by pandas.
        index = [u"Q24980", u"Q25043", u"Q24953", u"Q16403"]
        dataframe = pandas.DataFrame(records, index=pandas.Index(index, name="wikidata_id"))
>       job = client.load_table_from_dataframe(dataframe, table_ref, location="US")
docs/snippets.py:2779:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
google/cloud/bigquery/client.py:1535: in load_table_from_dataframe
    dataframe, job_config.schema
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <google.cloud.bigquery.job.LoadJobConfig object at 0x7f99632d7d90>
value = None
    @schema.setter
    def schema(self, value):
>       if not all(hasattr(field, "to_api_repr") for field in value):
E       TypeError: 'NoneType' object is not iterable
google/cloud/bigquery/job.py:1163: TypeError
=============================== warnings summary ===============================
docs/snippets.py::test_load_table_from_dataframe[pyarrow]
  /tmpfs/src/github/google-cloud-python/bigquery/google/cloud/bigquery/_pandas_helpers.py:229: UserWarning: Unable to determine type of column 'title'.
    warnings.warn("Unable to determine type of column '{}'.".format(column))
```

**Comment by tswast** [comment]
On second thought, since the snippets show writing the indexes of a DataFrame, I can't actually update them until https://github.com/googleapis/google-cloud-python/issues/5572#issuecomment-524012299 is fixed. I'll work on that issue next.

File: bigquery/tests/system.py (hunk 2 above)
Comment (tswast) [reply bigquery/tests/system.py]: Good point. Done in ca303a6c17536164c445f40af38808319e868106.
